Update MNIST example w.r.t. KFP 2.5.0

roopak-rastogi opened the pull request:
Provide a clear and concise description of the content changes you're proposing. List all the changes you are making
to the content.

- Updated section ___ with ___.
- Added new subsection ___ under ___.
- Removed outdated information from ___.
- ...

Closes #...

> If there is no issue related to this PR, kindly create one first to describe the motivation behind these changes.

Checklist:

- [x] I have checked that my enhancements are not duplicates of existing content changes or additions.
- [x] I have tested the changes in a working environment to ensure they function as intended.
- [ ] I have followed the [style guide](https://github.com/HPEEzmeral/ezua-tutorials/blob/develop/CONTRIBUTING.md#style-guide)
      outlined in the contribution guidelines.

Reviewer's Tasks (for maintainers reviewing this PR):

- [ ] Verify that the tutorial functions correctly in a live environment.
- [ ] Verify that the updated content aligns with the [style guide](https://github.com/HPEEzmeral/ezua-tutorials/blob/develop/CONTRIBUTING.md#style-guide)
      in the contribution guidelines.
- [ ] Check for consistency, grammar, and clarity throughout the updated content.
- [ ] Check that the related GitHub issue is up-to-date.

CybervisionVolodymyrPogorelov commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -41,7 +41,8 @@
     "# Install required packages (Kubeflow Pipelines and Katib SDK).\n",
     "import sys\n",
-    "!{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0"
+    "# !{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0\n",
```
I think that we don't need to install kfp in the example itself. It should be available in the notebook by default.

roopak-rastogi replied:
Agreed. I'll remove it.

roopak-rastogi replied:
Done

umka1332 replied:
It actually was worse then we don't need to - it was installation of the previous version, so if user uncomment that, it could brake whole example.

p.s. it was also already covered by my point from the comment above.:

> Commenting code make sense only, if it is expected that user will uncomment it for any reason (like setting proxy), but old code should just be deleted.

CybervisionVolodymyrPogorelov reviewed (changes requested):
We need to use KFP that goes in the notebook by default

roopak-rastogi commented:
> 1. Copy-pasting is bad - either figure ou,t how to get volume name from volume operation object, or define a constant.
> 2. Passing parameters that are unused is bad, but maybe you will use that volume operation object.
> 3. Commenting code make sense only, if it is expected that user will uncomment it for any reason (like setting proxy), but old code should just be deleted.
> 4. Using of the notebooks image is a clever idea, but it lacks airgap prefix. @AyushSinha5588 has added it to notebook as env var in past. Also that image is too big for what you are trying to achieve here. I've just solved similar problem for an another use case - https://jira-pro.it.hpe.com:8443/browse/EZAF-5507. We might want to add a new image for that or reuse some of kfp images like kfp-launcher or kfp-diver.

Addressed all the remarks. 
For the point 2: Passing parameters that are unused is bad, but maybe you will use that volume operation object.
Currently, model_volume is being passed as an argument to the functions, which is not being utilized/referenced directly. Due to the open issue of accessing passed arguments value (as of now, accessing passed arguments value is not accessed, rather than it shows `{{channel:task=;name=name;type=String;}}` ), once the related issue is resolved in the future release, the value can be referenced directly from the arguments passed. Therefore, leaving the argument as it is.

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:
```diff
@@ -9,9 +9,9 @@ inputs:
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: Worker Spec,                 type: String, default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String, default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
 - {name: Tfjob Timeout Minutes,       type: Integer,    default: 1440,      description: 'Time in minutes to wait for the TFJob to complete.'}
```
Please fix indentation - just like it is done for Tfjob Timeout Minutes

roopak-rastogi replied:
Done.

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:
```diff
@@ -9,9 +9,9 @@ inputs:
 - {name: PS Spec,                     type: JsonObject, default: '{}',      description: 'TFJob ps replicaSpecs.'}
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: Worker Spec,                 type: String,     default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String,     default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String,     default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
```
`PS Spec` should probably be also changed.

umka1332 replied:
We are not using it, but the same is true for `Evaluator Spec`

roopak-rastogi replied:
Done

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -52,7 +52,8 @@
     "# Specify the name of the user mounted directory\n",
-    "user_mounted_dir_name = \"user\""
+    "user_mounted_dir_name = \"user\"\n",
+    "airgap_registry=\"lr1-bd-harbor-registry.mip.storage.hpecorp.net/develop/\""
```
We need to get it from env var

roopak-rastogi replied:
Added the airgap_registry from env in the base_image path

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -568,19 +575,18 @@
+    "        pvc_name=volume_str,\n",
     "        size=\"1Gi\",\n",
-    "        modes=dsl.VOLUME_MODE_RWO\n",
+    "        storage_class_name=\"<storage_class_name>\",\n",
```
We can use something like this instead of expecting user to know this and notice to change.
```
storage_class_name = !kubectl get pvc user-pvc -o jsonpath={.spec.storageClassName}
storage_class_name = storage_class_name[0]
```

roopak-rastogi replied:
Updated the code with the above code snippet.

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -39,9 +39,11 @@
    "source": [
     "# Uncomment & specify proxy value if applicable\n",
     "# %env https_proxy=http://<proxy_host>:<port>\n",
+    "# Comma separated list of IP address\n",
+    "# %env no_proxy=\"<host1>:<port1>,<host2>:<port2>,...\"\" \n",
```
there could be not only IPs, but also IPs with masks and text hostnames. And afaik, it can not contain ports contrrary to httss_proxy.

roopak-rastogi replied:
Updated the comment and also added list in the `no_proxy`  as an example to provide better clarity. If anything else is also needed, please update me.

akravacyber commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -39,9 +39,11 @@
     "# Install required packages (Kubeflow Pipelines and Katib SDK).\n",
     "import sys\n",
-    "!{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0"
+    "!{sys.executable} -m pip install kfp[kubernetes]"
```
Do we need to install it actually? If we do not install `kfp[kubernetes]`, then example will fail, right? Looks strange, I've thought that we install the whole kfp package. If example is not working without it, I'm ok to leave it here. But in future in notebooks we can install this in all notebook images I think

roopak-rastogi replied:
We can add it in the notebooks in future. Have tested it without installing kfp[kubernetes] it was failing as we are calling it for creating the PVC.

akravacyber replied:
@roopak-rastogi thank you, got it, okay. It is fine

umka1332 replied:
Now, due to issue with `storageClassName` we are creating pvc with `kubectl`, so `kfp.kubernetes` module is not needed any more.

akravacyber commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -568,19 +576,19 @@
+    "        pvc_name=volume_str,\n",
     "        size=\"1Gi\",\n",
-    "        modes=dsl.VOLUME_MODE_RWO\n",
+    "        storage_class_name=storage_class_name[0],\n",
```
Do we need to pass a storage class name? I think it is not required, and if we omit it here - default storage class will be used by kubernetes

roopak-rastogi replied:
Without storage_class_name parameter, PVC was getting created in `standard` storage_class, since this storage_class is not present with this name, therefore, PVC and the TF jobs were going in Pending state. Same was seen when we tried with empty ("") storage_class_name.

akravacyber replied:
Hmm, got it, okay, Than yes, we can leave it as is

akravacyber reviewed (commented):
Actually looks good, left some minor comments

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -372,10 +376,16 @@
+    "from kfp import components\n",
+    "@dsl.component(\n",
+    "    install_kfp_package = False,\n",
+    "    base_image = f\"{os.environ['AIRGAP_REGISTRY']}gcr.io/mapr-252711/kubeflow/python-kfp:3.7-2.7.0\",\n",
```
Specifying base image and disabling kfp install are not needed any more - we have merged PR, that allows package installation:

- https://github.hpe.com/hpe/ezaf/pull/2855
- https://github.com/mapr/private-manifests/pull/584
- and most impotent - https://github.com/mapr/private-kubeflow-pipelines/pull/32

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:
```diff
@@ -388,12 +396,23 @@
-    "    convert_katib_results_op = components.func_to_container_op(convert_katib_results)\n",
-    "    best_hp_op = convert_katib_results_op(katib_op.output)\n",
+    "    best_hp_op = convert_katib_results(katib_results = katib_op.output)\n",
+    "    # Setting the proxy and no_proxy to allow download of the required packages inside the component. Uncomment the below code, if proxies are specified above.\n",
```
Please remove or update the comment. In case of update please notice that it is not related to `set_caching_options` that sits just below it.

roopak-rastogi replied:
Updated the comment and moved the `set_caching_options` part before the comment.

umka1332 commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/pvc.yaml:
```diff
@@ -0,0 +1,10 @@
+   - ReadWriteOnce
+  resources:
+   requests:
+    storage: 1Gi
```
Please add a newline at the end of file.

roopak-rastogi replied:
Done

roopak-rastogi commented on Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:
```diff
@@ -8,10 +8,10 @@ inputs:
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: PS Spec,                     type: String,     default: '{}',      description: 'TFJob ps replicaSpecs.'}
+- {name: Worker Spec,                 type: String,     default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String,     default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String,     default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
```
Updated the file with the mentioned comment.
